File: Samples/CustomEffectSample/CustomNativeEffects/CpuBasedEffects/CustomGrayscaleEffect.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************
#include "pch.h"
#include "CustomGrayscaleEffect.h"
#include "CustomGrayscaleCpuWorker.h"

using namespace CustomNativeEffects;
using namespace Platform;
using namespace Lumia::Imaging;
using namespace Concurrency;
using namespace Lumia::Imaging::Adjustments;

CustomGrayscaleEffect::CustomGrayscaleEffect()
{
}

IImageProvider2^ CustomGrayscaleEffect::Clone()
{
	critical_section::scoped_lock lock(m_criticalSection);

	auto clone = ref new CustomGrayscaleEffect();
	clone->m_properties = m_properties;
	clone->m_source = m_source->Clone();
	return clone;
}

RenderOptions CustomGrayscaleEffect::SupportedRenderOptions::get()
{
	return RenderOptions::Cpu;
}

Workers::IImageWorker^ CustomGrayscaleEffect::CreateImageWorker(Workers::IImageWorkerRequest^ imageWorkerRequest)
{
	if(imageWorkerRequest->RenderOptions == RenderOptions::Cpu)
	{
		return ref new CustomGrayscaleCpuWorker(this);
	}

	return nullptr;
}

IImageProvider^ CustomGrayscaleEffect::Source::get()
{
	concurrency::critical_section::scoped_lock lock(m_criticalSection);
	return m_source;
}

void CustomGrayscaleEffect::Source::set(IImageProvider^ value)
{
	concurrency::critical_section::scoped_lock lock(m_criticalSection);
	m_source = safe_cast<IImageProvider2^>(value);
}

uint32 CustomGrayscaleEffect::SourceCount::get()
{
	return 1;
}

void CustomGrayscaleEffect::GetSources(Platform::WriteOnlyArray<IImageProvider2^>^ sources)
{
	if(SourceCount > sources->Length)
	{
		throw ref new Platform::InvalidArgumentException("sources");
	}

	concurrency::critical_section::scoped_lock lock(m_criticalSection);

	sources[0] = m_source;
}

void CustomGrayscaleEffect::SetSource(uint32 sourceIndex, IImageProvider2^ source)
{
	if(sourceIndex >= SourceCount)
	{
		throw ref new Platform::InvalidArgumentException("sourceIndex");
	}

	concurrency::critical_section::scoped_lock lock(m_criticalSection);

	m_source = source;
}

File: Samples/CustomEffectSample/CustomEffects/CustomGrayscaleEffect.cs

//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************
using Lumia.Imaging;
using Lumia.Imaging.Workers;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Reflection;
using Windows.Storage.Streams;

namespace CustomEffects
{
    public class CustomGrayscaleEffect : EffectBase
    {
        public IBuffer PixelShader { get; set; }
        public CustomGrayscaleEffect()
        {
            LoadPixelShader();
        }

        public override RenderOptions SupportedRenderOptions
        {
            get { return RenderOptions.Mixed; }
        }

        public void LoadPixelShader()
        {
            var assembly = typeof(CustomGrayscaleEffect).GetTypeInfo().Assembly;

            Stream stream = assembly.GetManifestResourceStream("CustomEffects.Grayscale.pso");
            var streamReader = new MemoryStream();
            stream.CopyTo(streamReader);
            PixelShader = streamReader.ToArray().AsBuffer();
        }

        public override IImageProvider2 Clone()
        {
            var effect = new CustomGrayscaleEffect();
            effect.Source = ((IImageProvider2)Source).Clone();

            return effect;
        }

        public override IImageWorker CreateImageWorker(IImageWorkerRequest imageWorkerRequest)
        {
            if (imageWorkerRequest.RenderOptions == RenderOptions.Cpu)
            {
                return new CustomGrayscaleCpuWorker();
            }
            else if (imageWorkerRequest.RenderOptions == RenderOptions.Gpu)
            {
                return new CustomGrayscaleDirect2DWorker(this);
            }

            return null;
        }
    }
}
